### PR TITLE
[Core] Object spilling - Changed (remote) object restoration request to a long-poll, with exponentially backed off retries on failures.

### LINKED ISSUE #14646: [Core] Object spilling - Move object restoration requests to long-poll
The pull manager restoration requests currently retry without waiting for prior restoration requests to succeed or fail, resulting in a raylet getting spammed with restoration requests if those requests are failing or taking a long time to service.

We should try to move these restoration RPCs to a long-poll, where the RPC only returns when the restoration has either succeeded, in which case we try to pull the object, or failed, in which case we can either (a) immediately retry the RPC, or (b) rely on the retry timer to eventually retry the RPC (with exponential back off).

### PR DESCRIPTION
**Pull request by clarkzinzow**
## Why are these changes needed?

The pull manager restoration requests currently retry without waiting for prior restoration requests to succeed or fail, resulting in a raylet getting spammed with restoration requests if those requests are failing or taking a long time to service.

This PR moves these restoration RPCs to a long-poll, where the RPC only returns when the restoration has either succeeded, in which case we try to pull the object, or failed, in which case we update the retry timer for this pull so the pull manager eventually retries the RPC (with exponential back off).

This PR also includes some drivebys that improve error messages.

## TODOs

- [x] Update local object manager tests.
- [ ] Test this manually on spamming failure case, once the wheel is built.

## Related issue number

Closes #14646 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/object_manager/pull_manager.cc [hunk 1]
```diff
@@ -350,24 +355,37 @@ void PullManager::TryToMakeObjectLocal(const ObjectID &object_id) {
   // sending RPCs.
   if (!request.spilled_url.empty()) {
     const auto spilled_node_id = request.spilled_node_id;
+    // Set a flag on the request so we don't send duplicate restoration requests.
+    request.is_being_restored = true;
     restore_spilled_object_(
         object_id, request.spilled_url, spilled_node_id,
-        [this, object_id, spilled_node_id](const ray::Status &status) {
+        [this, object_id, &request, spilled_node_id](const ray::Status &status) {
```
Comment (clarkzinzow) [inline src/ray/object_manager/pull_manager.cc:362]: We may want to look up the pull request within the callback instead of capturing it by reference here, although I don't think that we're in danger of that captured reference getting invalidated?

File: src/ray/raylet/local_object_manager.cc [hunk 2]
```diff
@@ -372,34 +369,34 @@ void LocalObjectManager::AsyncRestoreSpilledObject(
...
+      callback(Status::PendingRequiredData("Object hasn't been spilled yet."));
     }
     return;
   }
 
   RAY_CHECK(objects_pending_restore_.emplace(object_id).second)
       << "Object dedupe wasn't done properly. Please report if you see this issue.";
-  io_worker_pool_.PopRestoreWorker([this, object_id, object_url, callback](
+  io_worker_pool_.PopRestoreWorker([this, object_id, object_url,
+                                    callback = std::move(callback)](
```
Comment (clarkzinzow) [inline src/ray/raylet/local_object_manager.cc:381]: C++14 generalized lambda capture with move captures! 😍

File: src/ray/object_manager/pull_manager.cc [hunk 3]
```diff
@@ -350,24 +355,37 @@ void PullManager::TryToMakeObjectLocal(const ObjectID &object_id) {
...
+                  << "Object restoration failed and the object " << object_id
+                  << " could not be found on any other nodes. This can happen "
+                     "if the location where the object was spilled is "
+                     "unreachable. This job may hang if the object is "
+                     "permanently unreachable. Please check the logs of node "
+                  << node_id_with_issue
+                  << " and the following status: " << status.ToString();
+            }
+            // Object restoration failed; we'll retry at a later time.
+            UpdateRetryTimer(request);
```
Comment (clarkzinzow) [inline src/ray/object_manager/pull_manager.cc:383]: Outside of the remote restoration requests, this also adds an exponential backoff for retrying when the object hasn't been spilled yet. I think that this is desirable, but I thought I'd call that out here in case that's controversial!

File: src/ray/object_manager/pull_manager.cc (hunk 1 above)
Comment (ericl) [reply src/ray/object_manager/pull_manager.cc:362]: It could be dangerous during destruction, I think you should change it to lookup.

File: src/ray/raylet/local_object_manager.cc [hunk 4]
```diff
@@ -233,15 +234,15 @@ void LocalObjectManager::SpillObjectsInternal(
               io_worker_pool_.PushSpillWorker(io_worker);
               if (!status.ok()) {
                 for (const auto &object_id : objects_to_spill) {
+                  RAY_LOG(ERROR) << "Failed to send object spilling request for object "
```
Comment (ericl) [inline src/ray/raylet/local_object_manager.cc:237]: Please revert this to the previous error message; this is too verbose.

File: src/ray/raylet/node_manager.cc [hunk 5]
```diff
@@ -2202,11 +2203,16 @@ void NodeManager::PublishInfeasibleTaskError(const Task &task) const {
 }
 
 void NodeManager::SendSpilledObjectRestorationRequestToRemoteNode(
-    const ObjectID &object_id, const std::string &spilled_url, const NodeID &node_id) {
+    const ObjectID &object_id, const std::string &spilled_url, const NodeID &node_id,
+    std::function<void(const ray::Status &)> callback) {
   // Fetch from a remote node.
   if (!remote_node_manager_addresses_.contains(node_id)) {
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc:2209]: Can you make the address a required argument of this method? That way we can remove this entire error path and make it a problem for the pull manager.

File: src/ray/raylet/node_manager.cc [hunk 6]
```diff
@@ -2202,11 +2203,16 @@ void NodeManager::PublishInfeasibleTaskError(const Task &task) const {
...
-    // It is possible the new node information is not received at this point.
-    // In this case, the PullManager will handle retry, so we just return.
+    // It is possible the new node information is not received at this point. The caller
+    // should retry at a later time.
+    if (callback) {
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc:2212]: Remove this error path per above

File: src/ray/raylet/node_manager.cc [hunk 7]
```diff
@@ -2216,12 +2222,17 @@ void NodeManager::SendSpilledObjectRestorationRequestToRemoteNode(
...
+      [object_id, node_id, callback = std::move(callback)](
+          const ray::Status &status, const rpc::RestoreSpilledObjectReply &r) {
         if (!status.ok()) {
-          RAY_LOG(WARNING) << "Failed to send a spilled object restoration request to a "
-                              "remote node. This request will be retried. Error message: "
+          RAY_LOG(WARNING) << "Failed to send a spilled object restoration request for "
+                           << object_id << " to a remote node " << node_id
+                           << ". This request will be retried. Error message: "
                            << status.ToString();
         }
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc:2232]: I don't think we need this log warn since we are passing the status to the callback, consider removing it.

**Review by ericl — CHANGES_REQUESTED** [review]
Couple other high level comments.
1. Don't you need to fix NodeManager::HandleRestoreSpilledObject() to reply asynchronously as well? Otherwise all this PR does is add the exponential stall even in the non-failure path.
2. Unit test for PullManager?

File: src/ray/raylet/local_object_manager.cc (hunk 4 above)
Comment (clarkzinzow) [reply src/ray/raylet/local_object_manager.cc:237]: As someone that has had to dig through logs to debug object spilling bugs, I strongly disagree. Knowing that an object spilling request failed without knowing which object failed or which IO worker was the target leaves you pretty blind, since you want know (1) what object to look for in the debug logs to determine if e.g. it was unexpectedly evicted, or (2) which IO worker logs to look at.

Comment (clarkzinzow) [reply src/ray/raylet/local_object_manager.cc:237]: I'd be fine with consolidating this into a single error message of `"Failed to send object spilling request for objects " << human_readable_list(objects_to_spill) << " to IO worker " << io_worker->WorkerId() << ": " << status.ToString();` if you're concerned about logging a message for every object. We already do this kind of logging once for every object in a bundle for e.g. activating and deactivating pull requests, but those tend to be debug logs.

Comment (ericl) [reply src/ray/raylet/local_object_manager.cc:237]: Keep in mind that all ERROR level logs are sent directly to the driver. If it's at all a normal condition that can occur during e.g., node failure, we should avoid excessive ERROR level logs.

File: src/ray/object_manager/pull_manager.cc (hunk 1 above)
Comment (clarkzinzow) [reply src/ray/object_manager/pull_manager.cc:362]: I _think_ that the thread that runs this handler should always run in a thread that's joined by the thread that created the pull manager, so the handler should never run after the pull requests table has been destructed, and IIUC we aren't at risk of this particular pull request being deleted or the reference being invalidated by concurrent table mutation...

But, since this isn't a super critical path, I think that it'd be better to change it to a lookup for the sake of simplicity and guarding against future bugs. 👍

File: src/ray/raylet/local_object_manager.cc (hunk 4 above)
Comment (clarkzinzow) [reply src/ray/raylet/local_object_manager.cc:237]: Would the consolidated single error message be tenable? I still think that erring on the side of verbosity here is better for the sake of those that will want to debug these kind of failures, unless we're expecting this to happen very frequently fail at large scale.

Comment (rkooo567) [reply src/ray/raylet/local_object_manager.cc:237]: Most of the users cannot debug this properly with this error message. I also prefer to make it less verbose. Having this message for each object will spam users' drivers and rather will confuse them.

Comment (clarkzinzow) [reply src/ray/raylet/local_object_manager.cc:237]: @rkooo567 What about the single error message that still lists each object, possibly truncating at a certain length? It will still be a single log record per failed spilling but will make it a lot easier for Ray devs to debug. In general, when logging the failure of a remote operation on a resource, including the resource and the identity of the remote node in the log record is considered standard practice, so I'm pretty bullish on us including at least one object ID and the IO worker ID in this log record.

Comment (clarkzinzow) [reply src/ray/raylet/local_object_manager.cc]: @rkooo567 @ericl I went with the single consolidated error message, here's an example output from running the local object manager tests:
```
[2021-03-16 17:53:18,193 E 12 12] local_object_manager.cc:247: Failed to send object spilling request for objects b050ebf366a840ea01b41d41a8e894f6b9c1d1597313212300000000 to IO worker d9f7a41d5dbe9926b2bfe43e72a25e4f639b8bc833c36eee633aab5c: IOError: error
```

**Comment by rkooo567** [comment]
I will review once Eric's initial comments are addressed. Please remove the author action required once they are all addressed!

**Review by ericl — APPROVED** [review]
LGTM

File: src/ray/raylet/local_object_manager.cc [hunk 8]
```diff
@@ -239,9 +240,13 @@ void LocalObjectManager::SpillObjectsInternal(
...
                   objects_pending_spill_.erase(it);
                 }
-
-                RAY_LOG(ERROR) << "Failed to send object spilling request: "
-                               << status.ToString();
+                std::ostringstream object_id_stream;
+                std::copy(objects_to_spill.begin(), objects_to_spill.end() - 1,
+                          std::ostream_iterator<ObjectID>(object_id_stream, ", "));
+                object_id_stream << objects_to_spill.back();
+                RAY_LOG(ERROR) << "Failed to send object spilling request for objects "
```
Comment (rkooo567) [inline src/ray/raylet/local_object_manager.cc:247]: Actually instead of this error message, why don't we just tell users which log files to look at? (the io worker log files that tried spilling objects).

File: src/ray/object_manager/pull_manager.cc [hunk 9]
```diff
@@ -326,6 +326,11 @@ void PullManager::TryToMakeObjectLocal(const ObjectID &object_id) {
   if (request.next_pull_time > get_time_()) {
     return;
   }
+  if (request.is_being_restored) {
```
Comment (rkooo567) [inline src/ray/object_manager/pull_manager.cc:329]: We still have 10ms interval retry in this case? And the only difference is we don't send a RPC right?

File: src/ray/object_manager/pull_manager.cc [hunk 10]
```diff
@@ -350,24 +355,44 @@ void PullManager::TryToMakeObjectLocal(const ObjectID &object_id) {
...
   if (!request.spilled_url.empty()) {
     const auto spilled_node_id = request.spilled_node_id;
+    // Set a flag on the request so we don't send duplicate restoration requests.
+    request.is_being_restored = true;
     restore_spilled_object_(
         object_id, request.spilled_url, spilled_node_id,
         [this, object_id, spilled_node_id](const ray::Status &status) {
+          auto it = object_pull_requests_.find(object_id);
+          if (it != object_pull_requests_.end()) {
+            // The pull request may no longer exist, e.g. it could have been cancelled.
```
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply src/ray/object_manager/pull_manager.cc:365]: When the actual object pull is cancelled while the restoration request is underway. I did observe this happening while running tests btw.

File: src/ray/raylet/local_object_manager.cc (hunk 8 above)
Comment (clarkzinzow) [reply src/ray/raylet/local_object_manager.cc:247]: I added the IO worker ID to the log message for that reason, are you talking about adding a more explicit directive for the user to go look at those logs, for users that don't know that those logs exist or that the worker ID is included in the log file name?

File: src/ray/raylet/local_object_manager.cc [hunk 11]
```diff
@@ -347,7 +352,9 @@ void LocalObjectManager::AsyncRestoreSpilledObject(
     const ObjectID &object_id, const std::string &object_url, const NodeID &node_id,
     std::function<void(const ray::Status &)> callback) {
   if (objects_pending_restore_.count(object_id) > 0) {
-    // If the same object is restoring, we dedup here.
+    // If the same object is restoring, we dedup here. We don't invoke the callback in
```
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply src/ray/raylet/local_object_manager.cc:355]: That's a good point, this should never be reached AFAICT.

File: src/ray/raylet/node_manager.cc [hunk 12]
```diff
@@ -2233,11 +2238,10 @@ void NodeManager::SendSpilledObjectRestorationRequestToRemoteNode(
...
   raylet_client->RestoreSpilledObject(
       object_id, spilled_url, node_id,
-      [](const ray::Status &status, const rpc::RestoreSpilledObjectReply &r) {
-        if (!status.ok()) {
-          RAY_LOG(WARNING) << "Failed to send a spilled object restoration request to a "
-                              "remote node. This request will be retried. Error message: "
-                           << status.ToString();
+      [callback = std::move(callback)](const ray::Status &status,
+                                       const rpc::RestoreSpilledObjectReply &r) {
+        if (callback) {
```
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply src/ray/raylet/node_manager.cc:2243]: I was going to ask about that check, did there used to be a path that didn't provide a callback? And if there is no such path anymore, do we even need the check? Why not just assume that a non-nullptr callback was provided, as we do with other callbacks?

File: src/ray/object_manager/pull_manager.cc (hunk 9 above)
Comment (clarkzinzow) [reply src/ray/object_manager/pull_manager.cc:329]: Yes, that is correct, this object won't be removed from the active pull request table while the restoration request is pending, but will instead short-circuit here on each retry timer tick.

File: src/ray/raylet/local_object_manager.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply src/ray/raylet/local_object_manager.cc:355]: Hmm actually @rkooo567, I hadn't thought of this, but I don't think that we're handling concurrent restoration requests from different nodes for the same object correctly. Consider this case (which I think can happen):

1. Node 1 spills object.
2. Node 2 issues remote restoration request to node 1.
3. Concurrently, node 1 issues a local restoration request.
4. The callbacks for one of these restoration requests will not be invoked due to deduping (with the current implementation), causing one of these pulls to never succeed.

With your suggestion (adding a RAY_CHECK here), one of those restoration requests would trip that check and cause a crash.

Am I right that this can happen? If so, we'd probably want restoration requests from different nodes to join on a single restoration that's processed, where each callback is invoked once the restoration is done, right?

File: src/ray/raylet/node_manager.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (rkooo567) [reply src/ray/raylet/node_manager.cc]: I think it is always safe to have a nullptr check because if (let's say by some mistakes) it tries to call nullptr, components silently die.

File: src/ray/raylet/local_object_manager.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (rkooo567) [reply src/ray/raylet/local_object_manager.cc]: Followed up on slack thread.

File: src/ray/raylet/local_object_manager.cc (hunk 8 above)
Comment (rkooo567) [reply src/ray/raylet/local_object_manager.cc:243]: Yes! because we are the only people who know how to debug when there is a worker id.

**Comment by rkooo567** [comment]
Fixing the concurrent restore + warning for many restoration requests in flight (I asked about it in the slack) are last requirements. If they are done, it LGTM

File: src/ray/raylet/local_object_manager.cc [hunk 13]
```diff
@@ -372,34 +360,40 @@ void LocalObjectManager::AsyncRestoreSpilledObject(
...
-  io_worker_pool_.PopRestoreWorker([this, object_id, object_url, callback](
+  auto it = objects_pending_restore_.find(object_id);
+  if (it != objects_pending_restore_.end()) {
+    // Restoration is already in progress, join that restoration by adding this request
+    // to the set of callbacks for that restoration.
+    it->second.push_back(callback);
+    return;
+  }
+  // Restoration is not yet in progress, so we proceed with starting a restoration.
+  it = objects_pending_restore_
```
Comment (rkooo567) [inline src/ray/raylet/local_object_manager.cc:375]: Why did you remove the check here?

File: src/ray/object_manager/pull_manager.cc [hunk 14]
```diff
@@ -350,24 +355,44 @@ void PullManager::TryToMakeObjectLocal(const ObjectID &object_id) {
...
+            it->second.is_being_restored = false;
```
Comment (rkooo567) [inline src/ray/object_manager/pull_manager.cc:366]: Is it guaranteed that after all callbacks are invoked after restoration, the pull requests are cancelled?

Comment (rkooo567) [reply src/ray/object_manager/pull_manager.cc:366]: Otherwise it is possible `is_being_restored` is still true after all of callbacks are completed right?

Comment (clarkzinzow) [reply src/ray/object_manager/pull_manager.cc:366]: >Is it guaranteed that after all callbacks are invoked after restoration, the pull requests are cancelled?

The pull request will be cancelled by higher-level code around task execution or fulfilling `ray.get()` or `ray.wait()`, but we don't want the pull requests to be cancelled right after restoration since we still need to pull the object.

>Otherwise it is possible is_being_restored is still true after all of callbacks are completed right?

There should be at most one `restore_spilled_object_` call, and therefore at most one callback, per pull request (due to our deduping on `request.is_being_restored`); after this callback executes, `request.is_being_restored` will be `false`, so the next time `TryToMakeObjectLocal` executes, it will pull the object.So I don't think that there's any way for `request.is_being_restored` to be true after the callback has completed.

Comment (clarkzinzow) [reply src/ray/object_manager/pull_manager.cc:366]: That did just make me think of something though... Upon a successful restoration, should we manually add the `spilled_node_id` to the locations for this request to be sure that the next `TryToMakeObjectLocal` won't trigger an unnecessary restoration request due to the object location subscription having not yet updated our view of the object locations yet?

File: src/ray/raylet/local_object_manager.cc (hunk 13 above)
Comment (clarkzinzow) [reply src/ray/raylet/local_object_manager.cc:378]: Because we short-circuit directly above this if an entry for `object_id` already exists in the map, so this should only be reached if there was was no entry for `object_id`. Besides, we're deduping in the pull manager now instead of here; multiple entries for `object_id` are supported due to us wanting to support multiple concurrent remote restoration requests.

File: src/ray/object_manager/pull_manager.cc (hunk 14 above)
Comment (clarkzinzow) [reply src/ray/object_manager/pull_manager.cc:366]: @rkooo567 I implemented what I think is more ideal behavior here, let me know if I'm missing something (if so I can revert that commit): https://github.com/ray-project/ray/pull/14660/commits/bf4a0df4fca53c3a3ab1bfbaac9a603318793792

Comment (rkooo567) [reply src/ray/object_manager/pull_manager.cc:366]: Sorry for being late! The current fix without this change is matching the previous behavior right (so this issue also happened before too)? Actually, can we just revert this and leave as TODO? I'd like to avoid adding more logic until we do refactoring.

**Review by rkooo567 — APPROVED** [review]
https://github.com/ray-project/ray/pull/14660#discussion_r597867928 => This is the last comment! Let's leave it out this logic for now and add TODO until we finish the refactoring. I think if the issue happens, it is pretty easy to discover (consumed bytes will be lower than restored bytes because we will do unnecessary restoration).

File: src/ray/object_manager/pull_manager.cc (hunk 14 above)
Comment (clarkzinzow) [reply src/ray/object_manager/pull_manager.cc:366]: No worries! And the previous behavior was to infinitely retry restoration requests until the object location subscription returned the restored node location; this PR is supposed to prevent unnecessary restoration retries, so I think that this fix is required to meet that criteria.

Comment (rkooo567) [reply src/ray/object_manager/pull_manager.cc:366]: Can we just do this later? I really want to avoid adding more logic right now as this path should work very reliably in the short term (for testing). More logic is more prone to error. Let's add this logic once things are easier to understand after refactoring.

Comment (rkooo567) [reply src/ray/object_manager/pull_manager.cc:366]: I think the current fix is already an improvement that it will avoid unnecessary RPCs a lot.

Comment (clarkzinzow) [reply src/ray/object_manager/pull_manager.cc:366]: I don't really get that reasoning, this new logic is fixing what I believe to be an error in the long-polling protocol, which is definitely within the scope of this PR. When the restoration request successfully completes, we know that the object is in the object store on the spilled node but the object location subscription may be lagging behind the restoration RPC, so we try to add it to the object locations. This is equivalent to an object location change, so we try to make the object local.

If I'm mistaken in this being an error in the long-polling protocol, then lmk and I'll revert it, but I don't think that we should be resistant to new logic if it fixes an error.

Comment (rkooo567) [reply src/ray/object_manager/pull_manager.cc:366]: Discussed offline. @clarkzinzow mentioned this was necessary due to the new changes, so we will just keep it here.

**Review by rkooo567 — APPROVED** [review]
Feel free to merge it when tests pass

**Comment by ericl** [comment]
I'm still working on this, so there might be some things I am missing, but this PR might be a cleaner approach: https://github.com/ray-project/ray/pull/14821

Please let me know if you have any suggestions!
